### ISSUE TITLE
GOSS benchmarking

### ISSUE THREAD

**Issue by Laurae2** [Microsoft/LightGBM#761]
@guolinke Is GOSS finalized enough to start some larger benchmarking on it?

Bosch dataset, 63 leaves, 40 threads, reported 500 iterations for timing:

| Format | 0 = NA | Booster | AUC | Best Iter | Time |
| --- | --- | --- | ---: | ---: | ---: |
| Sparse | FALSE | GBT | 0.7095987 | 140 | 135683.676 |
| Sparse | FALSE | GOSS | 0.7125379 | 110 | 96754.299 |
| Sparse | TRUE | GBT | 0.7097566 | 118 | 132794.350 |
| Sparse | TRUE | GOSS | 0.7103012 | 136 | 98313.085 |

So far it seems to provide promising results (both in speed and performance), not sure how it will behave with different parameters.

**Comment by guolinke** [comment]
@Laurae2 
The bottleneck of GOSS is it needs to prediction on full dataset on every iteration. 
We are working on the speed-up of prediction.

**Comment by Laurae2** [comment]
@guolinke Is it at prediction time or during training?

Also, looking at GOSS C++ code, I think the following about it:

* On noisy datasets (like Bosch), GOSS will improve the metric performance
* On synthetic datasets (like Higgs), GOSS will behave poorly and decrease the metric performance

Do you think this is right? I am starting benchmarks on LightGBM to check whether this is right or not.

**Comment by guolinke** [comment]
@Laurae2 it is in training. 
on synthetic datasets, most data have similar weights (gradients), so using GOSS may have loss. However, you can adjust top_rate and other_rate to let GOSS working in synthetic datasets as well.

**Comment by Laurae2** [comment]
@guolinke I'm seeing how slow GOSS is on my custom URL reputation dataset (the 2.25M xx 20K massive dataset).

BTW, it spends most of the time predicting in singlethread?

![image](https://user-images.githubusercontent.com/9083669/28996508-27d1e980-7a02-11e7-9e3f-f964562601e6.png)

**Comment by guolinke** [comment]
@Laurae2 yeah, it is slow due to prediction, especially when #feature is large. Actually, the prediction is done by multi-thread, But it has many branching (if-else), which causes the low usage per thread. 

However, another possible reason is you are using valid data(with metrics) during training .

**Comment by Laurae2** [comment]
@guolinke yes I am using a validation data during training. However I did not see it that slow with Bosch and Higgs even with validation data, it was 100% CPU usage.

**Comment by guolinke** [comment]
I see, so it will have two prediction costs, and both slow(due to large #feature).
And if you are using auc metric, it could be even slower since auc is also run by single thread. 

BTW, you can build LightGBM with "TIMETAG" flag, and it will output some timing statistics (with verbose=2). 
It will be better if you can provide the timing statistics, Thanks.

**Comment by Laurae2** [comment]
@guolinke How to build LightGBM with TIMETAG flag in R?

I am using fast AUC computation in R (https://github.com/Laurae2/R_benchmarking#area-under-the-curve-roc-benchmarks).

**Comment by guolinke** [comment]
Add it into cmake, like this https://github.com/Microsoft/LightGBM/blob/master/CMakeLists.txt#L33

**Comment by Laurae2** [comment]
It does not output timing statistics.

I added TIMETAG here: https://github.com/Laurae2/LightGBM/blob/test/CMakeLists.txt#L15

I also used `verbose = 2` but it prints nothing.

```r
lgb.unloader(wipe = TRUE)
library(lightgbm)
data(agaricus.train, package = "lightgbm")
train <- agaricus.train
dtrain <- lgb.Dataset(train$data, label = train$label)
params <- list(objective = "regression", metric = "l2", boosting = "goss")
model <- lgb.train(params,
                   dtrain,
                   100,
                   verbose = 2)
```

**Comment by Laurae2** [comment]
OK, I got them to appear but they appear only at the end.

10 trees

```
E:/build/LightGBM/Release/lightgbm.exe data=E:/benchmark_lot/data/reput_train_lgb.data verbose=9999 num_threads=8 learning_rate=0.25 max_depth=3 num_leaves=7 num_trees=10 metric=binary_logloss test=E:/benchmark_lot/data/reput_test_lgb.data app=binary boosting=goss
```

```
[LightGBM] [Info] GOSS::subset costs 31.101434
[LightGBM] [Info] GOSS::re_init_tree costs 60.731287
[LightGBM] [Info] GBDT::boosting costs 0.059216
[LightGBM] [Info] GBDT::train_score costs 0.080117
[LightGBM] [Info] GBDT::out_of_bag_score costs 0.000002
[LightGBM] [Info] GBDT::valid_score costs 0.006799
[LightGBM] [Info] GBDT::metric costs 0.006231
[LightGBM] [Info] GBDT::bagging costs 91.993470
[LightGBM] [Info] GBDT::sub_gradient costs 0.007929
[LightGBM] [Info] GBDT::tree costs 17.832810
[LightGBM] [Info] SerialTreeLearner::init_train costs 2.729572
[LightGBM] [Info] SerialTreeLearner::init_split costs 4.766684
[LightGBM] [Info] SerialTreeLearner::hist_build costs 9.566627
[LightGBM] [Info] SerialTreeLearner::find_split costs 0.603679
[LightGBM] [Info] SerialTreeLearner::split costs 0.136154
[LightGBM] [Info] SerialTreeLearner::ordered_bin costs 7.479034
```

You can compare with 3 trees, because at 5th tree it slows down considerably:

```
[LightGBM] [Info] GOSS::subset costs 0.000000
[LightGBM] [Info] GOSS::re_init_tree costs 0.000000
[LightGBM] [Info] GBDT::boosting costs 0.014698
[LightGBM] [Info] GBDT::train_score costs 0.023364
[LightGBM] [Info] GBDT::out_of_bag_score costs 0.000000
[LightGBM] [Info] GBDT::valid_score costs 0.001500
[LightGBM] [Info] GBDT::metric costs 0.001815
[LightGBM] [Info] GBDT::bagging costs 0.000002
[LightGBM] [Info] GBDT::sub_gradient costs 0.000000
[LightGBM] [Info] GBDT::tree costs 9.838733
[LightGBM] [Info] SerialTreeLearner::init_train costs 1.318383
[LightGBM] [Info] SerialTreeLearner::init_split costs 2.230926
[LightGBM] [Info] SerialTreeLearner::hist_build costs 6.098757
[LightGBM] [Info] SerialTreeLearner::find_split costs 0.173302
[LightGBM] [Info] SerialTreeLearner::split costs 0.015924
[LightGBM] [Info] SerialTreeLearner::ordered_bin costs 3.545788
```
